src/middlewared/middlewared/plugins/pool_/dataset_details.py

```diff
@@ -0,0 +1,221 @@
+        smb_shares = []
+        found = False
+        for smbshare in smbshares:
+            if smbshare['path'] == ds['mountpoint']:
+                # the share path is the actual dataset
+                found = True
+            else:
+                try:
+                    devid = os.stat(smbshare['path']).st_dev
+                except Exception:
+                    pass
+                else:
+                    found = devid in mntinfo and mntinfo[devid]['mount_source'] == ds['id']
+
+            if found:
+                smb_shares.append({
+                    'enabled': smbshare['enabled'],
+                    'path': smbshare['path'],
+                    'share_name': smbshare['name'],
+                })
+
+        return smb_shares
+
+    @private
+    def get_iscsi_shares(self, ds, iscsishares, mntinfo):
+        iscsi_shares = []
+        thick_provisioned = any((ds['reservation']['value'], ds['refreservation']['value']))
+        for share in iscsishares:
+            if share['extent']['type'] == 'DISK' and share['extent']['path'].removeprefix('zvol/') == ds['id']:
+                # we store extent information prefixed with `zvol/` (i.e. zvol/tank/zvol01).
+                iscsi_shares.append({
+                    'enabled': share['extent']['enabled'],
+                    'type': 'DISK',
+                    'path': f'/dev/{share["extent"]["path"]}',
+                    'thick_provisioned': thick_provisioned,
+                })
+            elif share['extent']['type'] == 'FILE':
+                # this isn't common but possible, you can share a "file"
+                # via iscsi which means it's not a dataset but a file inside
+                # a dataset so we need to find the source dataset for the file
+                try:
+                    devid = os.stat(share['extent']['path']).st_dev
+                except Exception:
+                    # this is a regular file so ignore exceptions for now
+                    pass
+                else:
+                    if devid in mntinfo and mntinfo[devid]['mount_source'] == ds['id']:
+                        iscsi_shares.append({
+                            'enabled': share['extent']['enabled'],
+                            'type': 'FILE',
+                            'path': share['extent']['path'],
+                            'thick_provisioned': thick_provisioned,
+                        })
+
+        return iscsi_shares
+
+    @private
+    def get_repl_tasks_count(self, ds, repltasks):
+        count = 0
+        for repl in filter(lambda x: x['direction'] == 'PUSH', repltasks):
+            # we only care about replication tasks that are configured to push
+            for src_ds in filter(lambda x: x == ds['id'], repl['source_datasets']):
+                count += 1
+
+        return count
+
+    @private
+    def get_cloudsync_tasks_count(self, ds, cldtasks):
+        return len([i for i in cldtasks if i['path'] == ds['mountpoint']])
```

Comment:
These seem to be handled differently than shares (we want exact path match), is that intentional?